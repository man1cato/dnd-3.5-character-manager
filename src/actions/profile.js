import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
