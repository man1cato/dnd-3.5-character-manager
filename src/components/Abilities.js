import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
