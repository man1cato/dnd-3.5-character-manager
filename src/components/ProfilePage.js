import React from 'react';
import {connect} from 'react-redux';

export const ProfilePage = (props) => (
    <div className="container container--body">
      <div className="grid grid--profile">
        <img className="grid--profile__img" src={props.iconUrl} />

        <div className="grid--profile__name">{props.name}</div>
        <div>{props.race}</div>
        <div>{props.gender}</div>
        <div>{props.class}</div>
        <div>{props.age} yrs</div>
        <div>{props.alignment}</div>
        <div>{props.height}</div>
        <div>{props.size}</div>
        <div>{props.weight} lbs</div>

        <h4 className="grid--profile__features-key">Languages</h4>
        <div className="grid--profile__features-value">{props.languages}</div>
        <h4 className="grid--profile__features-key">Feats</h4>
        <div className="grid--profile__features-value">{props.feats}</div>
        <h4 className="grid--profile__features-key">Special Abilities</h4>
        <div className="grid--profile__features-value">{props.specialAbilities}</div>
        <h4 className="grid--profile__features-key">School</h4>
        <div className="grid--profile__features-value">{props.school}</div>
        <h4 className="grid--profile__features-key">Prohibited Schools</h4>
        <div className="grid--profile__features-value">{props.prohibitedSchools}</div>
        <h4 className="grid--profile__features-key">Deity</h4>
        <div className="grid--profile__features-value">{props.deity}</div>
      </div>
    </div>
);

const mapStateToProps = (state) => ({
  ...state.profile
})

export default connect(mapStateToProps)(ProfilePage);
