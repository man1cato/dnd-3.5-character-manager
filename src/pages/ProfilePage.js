import React, { Fragment } from 'react'
import { connect } from 'react-redux'

import Feats from '../components/Feats'
import SpecialAbilities from '../components/SpecialAbilities'


export const ProfilePage = (props) => (
	<div className="container container--body">
		<div className="grid--profile">
			<img className="grid--profile__img" src={props.iconUrl} />

			<h4 className="grid--profile__name">{props.name}</h4>
			<div>{props.race.name}</div>
			<div>{props.gender}</div>
			<div>{props.jobClass.name}</div>
			<div>{props.age} yrs</div>
			<div>{props.alignment}</div>
			<div>{props.height}</div>
			<div>{props.race.size}</div>
			<div>{props.weight} lbs</div>

			<h4 className="grid--profile__features-key">Languages</h4>
			<div className="grid--profile__features-value">{props.languages.join(', ')}</div>

			<h4 className="grid--profile__features-key">Deity</h4>
			<div className="grid--profile__features-value">{props.deity}</div>

			{props.school && 
				<Fragment>
					<h4 className="grid--profile__features-key">School</h4>
					<div className="grid--profile__features-value">{props.school}</div>
				</Fragment>
			}

			{props.prohibitedSchools && 
				<Fragment>
					<h4 className="grid--profile__features-key">Prohibited Schools</h4>
					<div className="grid--profile__features-value">{props.prohibitedSchools.join(', ')}</div>
				</Fragment>
			}
				
			<h4 className="grid--profile__features-key">Special Abilities</h4>
			<div className="grid--profile__features-value">
				{props.specialAbilities &&
					<SpecialAbilities specialAbilityIds={props.specialAbilities} />
					||
					"None"
				}				
			</div> 

			<h4 className="grid--profile__features-key">Feats</h4>
			<div className="grid--profile__features-value">
				<Feats featIds={props.feats} />
			</div>

		</div>
	</div>
)


const mapStateToProps = ({profile, races, jobClasses}) => ({
	...profile,
	race: races[profile.race],
	jobClass: jobClasses[profile.jobClass]
})

export default connect(mapStateToProps)(ProfilePage)