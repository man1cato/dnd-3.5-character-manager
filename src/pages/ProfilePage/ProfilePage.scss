@import '../../styles/base/settings';


.ProfilePage__grid {
   display: grid;
   grid-template-columns: repeat(3, 1fr);
   grid-auto-flow: dense;
   grid-row-gap: $m-size;
   grid-column-gap: 1rem;
}

.ProfilePage__avatar {
   grid-area: 1 / 1 / span 5 / span 1;
   place-self: center;
   height: auto;
   width: 100%;
   padding: 5px;
   @media(min-width: $desktop-breakpoint) {
      width: 150px;
   }
}

.ProfilePage__name {
   font-weight: bold;
   grid-area: 1 / 2 / span 1 / span 2;
}

.ProfilePage__features {
   grid-column: 2 / span 2;
}