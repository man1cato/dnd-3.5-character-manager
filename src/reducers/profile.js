export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
