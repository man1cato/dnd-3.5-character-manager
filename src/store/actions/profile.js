import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
