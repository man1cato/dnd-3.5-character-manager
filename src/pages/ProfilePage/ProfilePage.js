import React from 'react'
import { connect } from 'react-redux'

import Feats from '../../components/Feats/Feats'
import SpecialAbilities from '../../components/SpecialAbilities/SpecialAbilities'
import './ProfilePage.scss'


export const ProfilePage = props => (
	<div className="container--body">
		<div className="ProfilePage__grid">
			<img className="ProfilePage__avatar" src={props.iconUrl} />

			<h4 className="ProfilePage__name">{props.name}</h4>
			<div>{props.race.name}</div>
			<div>{props.gender}</div>
			<div>{props.jobClass.name}</div>
			<div>{props.age} yrs</div>
			<div>{props.alignment}</div>
			<div>{props.height}</div>
			<div>{props.race.size}</div>
			<div>{props.weight} lbs</div>

			<h4 className="grid__col1">Languages</h4>
			<div className="ProfilePage__features">{props.languages.join(', ')}</div>

			<h4 className="grid__col1">Deity</h4>
			<div className="ProfilePage__features">{props.deity}</div>

			{props.school && 
				<>
					<h4 className="grid__col1">School</h4>
					<div className="ProfilePage__features">{props.school}</div>
				</>
			}

			{props.prohibitedSchools && 
				<>
					<h4 className="grid__col1">Prohibited Schools</h4>
					<div className="ProfilePage__features">{props.prohibitedSchools.join(', ')}</div>
				</>
			}
				
			<h4 className="grid__col1">Special Abilities</h4>
			<div className="ProfilePage__features">
				{props.specialAbilities &&
					<SpecialAbilities specialAbilityIds={props.specialAbilities} />
					||
					"None"
				}				
			</div> 

			<h4 className="grid__col1">Feats</h4>
			<div className="ProfilePage__features">
				<Feats featIds={props.feats} />
			</div>
		</div>
	</div>
)


const mapStateToProps = ({profile, races, jobClasses}) => ({
	...profile,
	race: races[profile.race],
	jobClass: jobClasses[profile.jobClass]
})

export default connect(mapStateToProps)(ProfilePage)