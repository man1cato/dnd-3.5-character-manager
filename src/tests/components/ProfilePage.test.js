import React from 'react';
import { shallow } from 'enzyme';

import { ProfilePage } from '../../components/ProfilePage';
import profile from '../fixtures/profile';

const profileData = {
	id: profile.id,
	...profile.fields
}


test('should render ProfilePage with profile data', () => {
	const wrapper = shallow(<ProfilePage {...profileData} />)
	expect(wrapper).toMatchSnapshot()
})
