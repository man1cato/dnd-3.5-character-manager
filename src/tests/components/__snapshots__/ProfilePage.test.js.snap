// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProfilePage with profile data 1`] = `
<div
  className="layout"
>
  <Connect(Header)
    pageTitle="Character Profile"
  />
  <div
    className="container container--body"
  >
    <div
      className="grid grid--profile"
    >
      <img
        className="grid--profile__img"
        src="images/elf.svg"
      />
      <h4
        className="grid--profile__name"
      >
        Istvyll Bannianea
      </h4>
      <div>
        Grey Elf
      </div>
      <div>
        Male
      </div>
      <div>
        Wizard
      </div>
      <div>
         yrs
      </div>
      <div>
        Lawful Evil
      </div>
      <div />
      <div />
      <div>
         lbs
      </div>
      <h4
        className="grid--profile__features-key"
      >
        Languages
      </h4>
      <div
        className="grid--profile__features-value"
      >
        Common, Elven, Draconic, Orc, Sylvan
      </div>
      <h4
        className="grid--profile__features-key"
      >
        Deity
      </h4>
      <div
        className="grid--profile__features-value"
      />
      <h4
        className="grid--profile__features-key"
      >
        School
      </h4>
      <div
        className="grid--profile__features-value"
      >
        Transmutation
      </div>
      <h4
        className="grid--profile__features-key"
      >
        Prohibited Schools
      </h4>
      <div
        className="grid--profile__features-value"
      >
        Illusion, Necromancy
      </div>
      <h4
        className="grid--profile__features-key"
      >
        Special Abilities
      </h4>
      <div
        className="grid--profile__features-value"
      >
        <Connect(SpecialAbilities)
          specialAbilityIds={
            Array [
              "rect54ikkC9SDLJ1x",
            ]
          }
        />
      </div>
      <h4
        className="grid--profile__features-key"
      >
        Feats
      </h4>
      <div
        className="grid--profile__features-value"
      >
        <Connect(Feats)
          featIds={
            Array [
              "recRjuAWe7oy6a7wY",
              "rec4KkfhKpetKIDqj",
              "recGkxwkqppy7EYmI",
            ]
          }
        />
      </div>
    </div>
  </div>
  <Connect(Footer) />
</div>
`;
